[dnf5] Enhance advisory query and remove enums

kontura opened the pull request:
I am conflicted about removing the enums, we are losing the enumeration of possible values but it seems a lot cleaner without them.

Advisory::get_references is most likely slower but it is no longer used in AdvisoryQuery.

kontura committed:
Remove ENUMs of Advisory and AdvisoryReference types

AdvisoryQuery needs to be able to accept the types as strings,
Libsolv stores them also as strings and we don't store them anywhere.

We were just converting from strings to enums and back.
Using just the strings is a lot simpler.

kontura committed:
Optimize AdvisoryQuery by using libsolv dataiterator for matching

Constructing Advisory* objects each iteration is expensive use libsolv
dataiterator instead, it can also do the string matching.

This brings the used approach closer to PackageQuery

lukash commented on include/libdnf/advisory/advisory_reference.hpp:
```diff
@@ -40,8 +38,8 @@ class AdvisoryReference {
-    /// @return Type of this reference.
-    Type get_type() const;
+    /// @return type of this reference as std::string.
+    std::string get_type() const;
```
Looks fine to me in this case to use strings instead of an enum. Though I'm not sure why `AdvisoryQuery` needs to accept the types as strings as you state in the commit description.

Also with this change I think we need to document the list of possible type values in the documentation. I think either in multiple places or (better) in a single place to which we link from those multiple places (i.e. for advisory type at least here in the `get_type()`'s docstring and then in `AdvisoryQuery::filter_type()`).

kontura replied:
> Looks fine to me in this case to use strings instead of an enum. Though I'm not sure why `AdvisoryQuery` needs to accept the types as strings as you state in the commit description.

I remember we discussed it with @j-mracek and I had a TODO for it but to be honest I kind of forgot the reasoning. :disappointed: 

> Also with this change I think we need to document the list of possible type values in the documentation. I think either in multiple places or (better) in a single place to which we link from those multiple places (i.e. for advisory type at least here in the `get_type()`'s docstring and then in `AdvisoryQuery::filter_type()`).

I have added a commit for it, adding it in multiple places for now because I am not sure how to make a link here. Do you mean just writing something like: "Possible types are described in the `get_type()` method."?

lukash replied:
I'm also not sure how to add a link, there's probably a way in Doxygen... I'm not sure myself if to link to a single place or list the types in multiple places. Listing in multiple places might actually be more straightforward and easier to read in the docs, so I think it's just fine the way you did it.

lukash commented on libdnf/advisory/advisory_query.cpp:
```diff
@@ -43,233 +43,166 @@ AdvisoryQuery::AdvisoryQuery(Base & base) : AdvisoryQuery(base.get_weak_ptr()) {
+        case libdnf::sack::QueryCmp::ICONTAINS:
+            return SEARCH_NOCASE | SEARCH_SUBSTRING;
+        default:
+            throw AdvisoryQuery::NotSupportedCmpType("Used unsupported CmpType");
```
I think this should just be an AssertionError, @jrohel @j-mracek do we agree in general? (I think we have a lot of these right?)

kontura replied:
Yea, this approach is used all over in PackageQuery as well. I think if we are going to change it we should do it all at once, perhaps in a separate PR.

j-mracek replied:
I agree, not supported cmb type or incorrect sack should be inherited from assertion error.

lukash commented on include/libdnf/advisory/advisory_query.hpp:
```diff
@@ -66,10 +66,6 @@ class AdvisoryQuery : public AdvisorySet {
-    AdvisoryQuery & filter_reference(
-        const std::vector<std::string> & reference_id,
-        std::vector<std::string> types,
-        sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
```
So this is going away? I see `filter_CVE` (btw. those capital letters seem to be against coding style?) and `filter_bug` (the name is not consistent with the type name which is `"bugzilla"`), there aas also the `"vendor"` type which can't be filtered now?

kontura replied:
Right, I have switched it back to the more general `filter_reference()` that takes a type as a param.

kontura committed:
Move and rename get_sorted_advisory_packages

Change the name to less misleading `get_advisory_packages_sorted_by_id`
and move the function to a private section of `AdvisorySet`.

Since we cannot test it there remove its test (and add a missing test
for `get_advisory_packages` which uses it).

kontura committed:
Add possible types for Advisory and AdvisoryReference

Since we use a std::string to encode the type list the possible values

kontura committed:
Add general `filter_reference` and remove specific bug/cve filters

kontura commented:
There is also a new commit that moves and renames `get_sorted_advisory_packages()` as we discussed on the meeting.

kontura commented:
Rebased to fix a conflict.

lukash commented on include/libdnf/advisory/advisory_collection.hpp:
```diff
@@ -63,6 +64,7 @@ class AdvisoryCollection {
 private:
     friend Advisory;
     friend AdvisoryQuery;
+    friend AdvisorySet;
```
Is this really needed? Seems it's not?

kontura replied:
Since I have moved the `get_advisory_packages_sorted_by_id` (`get_sorted_advisory_packages`) to `AdvisorySet` it is needed because it uses faster private version of `get_packages(..)` which appends the packages to an input vector (its much faster than returning a new vector for each collection and the coping the packages over).

lukash commented on include/libdnf/advisory/advisory_set.hpp:
```diff
@@ -127,7 +133,9 @@ class AdvisorySet {
 private:
     friend AdvisorySetIterator;
     friend class AdvisoryQuery;
+    friend class libdnf::rpm::PackageQuery;
```
This one also doesn't seem actually used in this commit.

kontura replied:
The `PackageQuery` is using `get_advisory_packages_sorted_by_id`.

lukash commented on libdnf/advisory/advisory_query.cpp:
```diff
@@ -301,22 +301,4 @@ std::vector<AdvisoryPackage> AdvisoryQuery::get_advisory_packages(
-        }
-    }
-
-    std::sort(out.begin(), out.end(), AdvisoryPackage::Impl::nevra_compare_lower_id);
```
But this might mean `AdvisoryQuery` could be dropped as `AdvisoryPackage`'s friend. We need to keep the friends lists as small as possible, it's already kind of crazy already...

kontura replied:
I am afraid `AdvisoryQuery` is still using `AdvisoryPackage::Impl::name_arch_compare_lower_id`.

lukash commented on libdnf/advisory/advisory_set.cpp:
```diff
@@ -137,5 +138,22 @@ BaseWeakPtr AdvisorySet::get_base() const {
+        }
+    }
+
+    std::sort(out.begin(), out.end(), AdvisoryPackage::Impl::nevra_compare_lower_id);
```
I think the `AdvisoryPackage::Impl::nevra_compare_lower_id` and friends could be pulled out of that class so that you don't need a friend and semantics-wise they are the same as for sack packages, right? Maybe a shared implementation? Templated, probably.

Treat this as a remark for now, no need to solve it here.

lukash commented:
Ok ok :slightly_smiling_face: thanks, merging.
